Post: deprecation of building without config file (#219)
readthedocs/blog: migrate-configuration-v2.rst

Here you have a minimal template configuration file that you can copy and paste to use in your project.
With this config file your project should build successfully now
and will keep working fine once we deploy the changes in September 2023.

.. code:: yaml

   # .readthedocs.yaml
   # Read the Docs configuration file
   # See https://docs.readthedocs.io/en/stable/config-file/v2.html for details

   # Required
   version: 2

   # Set the version of Python and other tools you might need
   build:
     os: ubuntu-22.04
     tools:
       python: "3.11"

   # Build documentation in the docs/ directory with Sphinx
   sphinx:
     configuration: docs/conf.py

Things to pay attention
-----------------------

It's worth to mention there are some small differences between the defaults used before and this configuration file.
If the first build fails after making these changes,
here is a checklist of items that you should pay attention to while writing your ``.readthedocs.yaml`` file:

* Operating system is changed from Ubuntu 18.04 to Ubuntu 22.04
* The new Ubuntu 22.04 may not have all the same system dependencies installed.

>   If you need an Ubuntu package that it's not installed in the system,

Reviewer: ```suggestion
  If you need an Ubuntu package that's not installed in the system,
```

  you can use
  `build.apt_packages <https://docs.readthedocs.io/en/stable/config-file/v2.html#build-apt-packages>`_
  to install it.

> * Python version is changed from 3.7 to 3.11.

Reviewer: ```suggestion
* The default Python version has changed from 3.7 to 3.11.
```

* All the configuration done from the dashboard
  (:menuselection:`Project Admin --> Advanced Settings --> Default settings`)

>   are ignored when using a configuration file and have to be migrated to this configuration file if you were using any of them.

Reviewer: ```suggestion
  is ignored when using a configuration file. These settings have to be migrated to your project's configuration file.
```

  * ``Documentation type`` is now defined via
    `sphinx <https://docs.readthedocs.io/en/stable/config-file/v2.html#sphinx>`_ or
    `mkdocs <https://docs.readthedocs.io/en/stable/config-file/v2.html#mkdocs>`_,

>     depending which is the documentation tool your project is using.

Reviewer: ```suggestion
    depending on which documentation tool your project is using.
```

  * ``Requirements file`` is now defined via
    `python.requirements <https://docs.readthedocs.io/en/stable/config-file/v2.html#requirements-file>`_.
  * ``Python Interpreter`` is now defined via
    `build.tools.python <https://docs.readthedocs.io/en/stable/config-file/v2.html#build-tools-python>`_.
  * ``Install Project`` is now defined via
    `python.install <https://docs.readthedocs.io/en/stable/config-file/v2.html#python-install>`_.
  * ``Use system packages`` *is deprecated* in favor of declaring all your dependencies via a requirements file using
    `python.requirements <https://docs.readthedocs.io/en/stable/config-file/v2.html#requirements-file>`_.
  * ``Python configuration file`` is now defined via
    `sphinx.configuration <https://docs.readthedocs.io/en/stable/config-file/v2.html#sphinx-configuration>`_.

> * Git submodules are not cloned.

Reviewer: ```suggestion
* Git submodules are not cloned automatically.
```

  You can tell Read the Docs to clone them by using
  `submodules.include <https://docs.readthedocs.io/en/stable/config-file/v2.html#submodules-include>`_
  config key.
* HTMLZip format is not built by default.
  In case you want to keep building it,
  you can use the config key
  `formats <https://docs.readthedocs.io/en/stable/config-file/v2.html#formats>`_.
* Node.js is not available by default.
  You can make it available by using
  `build.tools.nodejs <https://docs.readthedocs.io/en/stable/config-file/v2.html#build-tools-nodejs>`_
  config key.

> Starting with this suggested ``.readthedocs.yaml`` configuration file and taking into account the differences,

Reviewer: ```suggestion
Starting with this suggested ``.readthedocs.yaml`` configuration file and taking into account these differences,
```

> you will be able to update your project to the new requirements now and don't worry about this breaking change in the future.

Reviewer: ```suggestion
you should be able to define a working configuration file and avoid breaking changes in the future.
```

What's next?
------------

In case you have some extra time,

> we recommend you to pin your Python dependencies as well to avoid builds breaking by surprise when a new release of any of your dependencies is made.

Reviewer: ```suggestion
we recommend that you pin your project's Python dependencies as well.
This helps avoid surprise build errors when new packages are released.
```

You can read the guide `How to create reproducible builds <https://docs.readthedocs.io/en/stable/guides/reproducible-builds.html>`_ for this.

Contact us
----------

Get in touch with us `via our support`_

> and let us know if you are unable to use a configuration file for any reason.

Reviewer: ```suggestion
and let us know if you are having trouble using a configuration file for any reason.
```

.. post:: May 31, 2023
   :tags: builders, deprecation
   :author: Manuel
   :location: BCN
   :category: Changelog

Migrate your project to ``.readthedocs.yaml`` configuration file v2
===================================================================

We are announcing the **deprecation of building without a configuration file** (``.readtehdocs.yaml``)
together with the deprecation of configuration file v1 as well.

Read the Docs *will soon start requiring* a ``.readthedocs.yaml`` configuration file
for all the projects to build their documentation successfully.
We will stop creating a default configuration file *behind the scenes*, making assumptions for your projects

> and trying our best to make the builds succeed.

Reviewer 2: I think this should be removed, and then another set of paragraphs are expanded and moved to a "Background" section.

We plan to deploy these changes and **start failing builds without a configuration file on September 2023**.
This matches the period when ``urllib3`` `drops support for OpenSSL 1.1.1 <https://github.com/urllib3/urllib3/issues/2168>`_
which is OpenSSL version that Ubuntu 18.04 has installed and it's used when the project doesn't have a configuration file.

> The main reason behind this decision is it's pretty hard to maintain a set of defaults that are useful
> for the amount of users we have and the number of different tools they use to build documentation.
> Even when the defaults we chose were useful for a period of time,
> we found they were almost impossible to update to newer versions of dependencies
> because doing so would break a lot of projects that weren't using a configuration file
> and pinning their dependencies (e.g. OS, Python version, dependencies, etc).

Reviewer 2: ```suggestion
*Behind the scenes*, we used to assume a default configuration file for projects that didn't have one. However, a "one size fits all" default project setup isn't sustainable anymore, since it requires us to find a common ground for projects created in a period of over 10 years!

We would also much rather encourage all the positive aspects of having a configuration file. Maintaining your Read the Docs configuration together with your documentation allows you to version them together.

Moving forward, this change will make our platform much less complex. It means that we can introduce new features faster and that users will experience fewer issues due to the unstable nature of these one-size-fits-all legacy default setups.
```

I suggest to turn this into a "Background" section and move it to the end of the blog post. That way, they deprecation message and the practical parts can get the focus.

> If you have a project on Read the Docs that it's not using a ``.readthedocs.yaml`` configuration file,
> **we strongly recommend you to create one as soon as possible** to avoid breaking builds once we deploy the changes.

Reviewer 2: ```suggestion
Add your first .readthedocs.yaml

If you have a project on Read the Docs that is not using a ``.readthedocs.yaml`` configuration file,
**we strongly recommend you to create one as soon as possible** to avoid breaking builds once we deploy the changes.
```

> Here you have a minimal template configuration file that you can copy and paste to use in your project.
> With this config file your project should build successfully now
> and will keep working fine once we deploy the changes in September 2023.

Reviewer 2: ```suggestion
The following example contains a minimal template configuration that uses our default Sphinx builder.

If you configured your project as a Sphinx project using the Dashboard UI,
then this would have been the same configuration file outcome *behind the scenes*.
```

.. code:: yaml

   # .readthedocs.yaml
   # Read the Docs configuration file
   # See https://docs.readthedocs.io/en/stable/config-file/v2.html for details

   # Required
   version: 2

   # Set the version of Python and other tools you might need
   build:
     os: ubuntu-22.04
     tools:
       python: "3.11"

   # Build documentation in the docs/ directory with Sphinx
   sphinx:
     configuration: docs/conf.py

> It's worth to mention there are some small differences between the defaults used before and this configuration file.
> These differences may require some manual adjustment that you will figure if the first build after making these changes fails:

Reviewer 2: ```suggestion
It's worth mentioning that there differences between the defaults used before and this configuration file.

If the first build fails after making these changes,
here is your checklist of items that you should pay attention to, specially *if you have been using non-default Dashboard UI options*:
```

* Operating system is changed from Ubuntu 18.04 to Ubuntu 22.04
* The new Ubuntu 22.04 may not have all the same system dependencies installed.
  If you need an Ubuntu package that it's not installed in the system,
  you can use
  `build.apt_packages <https://docs.readthedocs.io/en/stable/config-file/v2.html#build-apt-packages>`_
  to install it.

> * Python version is changed from 3.7 to 3.11.

Reviewer 2: ```suggestion
* Python version is changed from 3.7 to 3.11.
* If you used any of the following settings in the website's :menuselection:`Project Admin --> Advanced Settings --> Default settings`, you should adjust the configuration file accordingly. 
  * Documentation type: If you didn't choose "Sphinx" here, please read ...
  * Requirements file: Please see...
  * ...
```

I think that this could need some more elaboration. Maybe a little list that says "For this UI field, you need to use this corresponding .readthedocs.yaml configuration key"

![image](https://github.com/readthedocs/blog/assets/374612/6f1be703-aa85-4833-b2ab-f2c28d1635dd)

Reviewer 2: Ah yes, the reason that I added the arrows in the above screenshot was that I thought a subsequent change would be to add documentation links to all of the help texts for these options, pointing to the .readthedocs.yaml key to use instead.

> * Git submodules are not clonned.

Reviewer 2: ```suggestion
* Git submodules are not cloned.
```

:heart:

  You can tell Read the Docs to clone them by using
  `submodules.include <https://docs.readthedocs.io/en/stable/config-file/v2.html#submodules-include>`_
  config key.
* HTMLZip format is not built by default.
  In case you want to keep building it,
  you can use the config key
  `formats <https://docs.readthedocs.io/en/stable/config-file/v2.html#formats>`_.
* Node.js is not available by default.
  You can make it available by using
  `build.tools.nodejs <https://docs.readthedocs.io/en/stable/config-file/v2.html#build-tools-nodejs>`_
  config key.

> Starting with this suggested ``.readthedocs.yaml`` configuration file and taking into account the differences,
> you will be able to update your project to the new requirements now and don't worry about this breaking change.

Reviewer 2: ```suggestion
If there is anything that we missed in this checklist or you have issues with these instructions, 
get in touch with us `via our support`_.

.. _via our support: https://readthedocs.org/support/
```

> 

Reviewer 2: I was thinking that maybe we should have something about what happens to the Project Admin?

```suggestion

Configuration via website UI
----------------------------

Our Project Admin currently contains a lot of options that will be deprecated.
You can continue to use the dashboard to understand your previous configuration.
However, many options will disappear from the Project Admin once we finally remove the support for projects without a ``.readthedocs.yaml`` file.

I added a configuration file and it worked! What's next?
--------------------------------------------------------

```

Finally, in case you have some extra time,
we recommend you to pin your Python dependencies as well to avoid builds breaking by surprise when a new release of any of your dependencies is made.
You can read the guide `How to create reproducible builds <https://docs.readthedocs.io/en/stable/guides/reproducible-builds.html>`_ for this.

Get in touch with us `via our support`_
and let us know if you are unable to use a configuration file for any reason.

> .. _via our support: https://readthedocs.org/support/

Reviewer 2: Suggestion to move up.
